In [47]:
import numpy as np
import pandas as pd
import scipy.stats as stats
import math

In [48]:
population_age=np.random.randint(18,75,1000)

In [49]:
population_age

array([58, 36, 32, 59, 45, 41, 50, 42, 45, 29, 43, 60, 68, 54, 53, 73, 66,
       38, 43, 59, 52, 27, 72, 60, 72, 18, 47, 68, 74, 22, 18, 47, 52, 30,
       29, 52, 50, 68, 41, 19, 64, 51, 36, 60, 43, 62, 55, 48, 67, 44, 26,
       56, 69, 58, 43, 63, 61, 60, 58, 60, 20, 31, 25, 51, 20, 68, 32, 19,
       67, 41, 34, 66, 32, 45, 73, 70, 33, 21, 50, 72, 59, 52, 28, 71, 73,
       67, 27, 37, 66, 64, 47, 37, 51, 34, 72, 70, 72, 26, 30, 36, 41, 74,
       74, 24, 19, 63, 63, 34, 54, 50, 62, 24, 43, 46, 28, 23, 40, 64, 18,
       46, 55, 34, 56, 70, 27, 55, 49, 34, 62, 52, 68, 24, 24, 44, 71, 49,
       27, 20, 25, 68, 29, 38, 54, 43, 71, 35, 47, 55, 66, 31, 55, 45, 62,
       20, 50, 22, 44, 69, 24, 46, 23, 56, 31, 42, 64, 43, 43, 73, 21, 55,
       47, 50, 49, 73, 39, 41, 69, 34, 30, 31, 61, 44, 40, 19, 59, 66, 44,
       40, 20, 21, 30, 41, 36, 66, 27, 72, 48, 32, 74, 48, 28, 74, 68, 54,
       27, 49, 25, 71, 44, 64, 29, 41, 50, 32, 32, 67, 46, 55, 36, 35, 65,
       35, 65, 33, 52, 58

In [50]:
# now let's take a sample of 100 ages randomly

sample_age=np.random.choice(population_age,100)
sample_age

array([52, 65, 32, 39, 32, 35, 69, 26, 31, 45, 33, 42, 66, 27, 19, 24, 43,
       30, 68, 19, 33, 58, 31, 50, 21, 70, 52, 62, 55, 74, 22, 63, 21, 42,
       32, 71, 68, 31, 32, 72, 56, 23, 56, 38, 74, 52, 67, 21, 52, 52, 22,
       65, 46, 46, 63, 36, 28, 57, 54, 54, 56, 22, 36, 62, 57, 38, 41, 40,
       55, 29, 62, 72, 65, 60, 37, 71, 32, 30, 58, 50, 27, 59, 26, 52, 61,
       41, 50, 35, 40, 68, 22, 52, 62, 43, 69, 44, 54, 25, 42, 69])

In [51]:
# now let's find out the mean of sample and population

print(np.mean(population_age))
print(np.mean(sample_age))

45.889
46.33


#### now here we get the mean of population and sample which are different, but how can we be able to know thaṭ the sample is a true representation of the population. By just looking at the mean we can't say that. 
### So here comes the idea of hypothesis testing:
####               The null hypothesis says that both sample and population are same means sample represents the population properly, means they don't have any difference
####               The alternate hypothesis says that they both are dofferent, means the sample doesn't represent the population

### So For that we need to do the basic t-test which is called as one sample T-test, because we are comparing the mean of the sample and the mean of population to find out whether the sample represents the population or not and also whether the NULL hypothesis is true or the Alternate hypothesis is true

#### now as we know that based on the significance value we can carry out the hypothesis testing, so let's take the significance value which is the p value as the conventional 0.05 which means that if the significance value goes less than the p-value or the confidence value of 5% then we will reject the NULL hypothesis and will accept the Alternate hypothesis

In [52]:
# let's import the 1-sample test

from scipy.stats import ttest_1samp

In [53]:
ttest_1samp(sample_age,np.mean(population_age))

Ttest_1sampResult(statistic=0.27213556217262636, pvalue=0.7860845338314219)

### No wlet's discuss about 2-sample t test

#### So in the 2-sample t test we basically compare 2 different samples to find out whether they are same or not.
#### So here the NULL hypothesis says that the 2 samples we get are from the same distribution but ALTERNATE hypothesis says that the 2 samples are from different distributions

In [54]:
# Let's take an example

ages_sample2=stats.poisson.rvs(loc=18,mu=40,size=100) 

In [63]:
ages_sample2

array([63, 58, 55, 57, 66, 51, 54, 47, 58, 43, 60, 70, 58, 56, 54, 57, 58,
       61, 57, 57, 60, 53, 76, 58, 68, 55, 63, 58, 74, 64, 64, 59, 56, 58,
       58, 67, 42, 59, 58, 62, 59, 55, 58, 72, 64, 56, 62, 55, 58, 51, 55,
       51, 62, 53, 66, 68, 63, 56, 51, 56, 47, 63, 56, 48, 54, 52, 48, 55,
       62, 60, 63, 63, 72, 69, 56, 60, 60, 69, 61, 59, 54, 48, 62, 60, 61,
       55, 68, 54, 64, 65, 51, 51, 61, 48, 61, 56, 54, 59, 65, 63])

In [56]:
sample_age

array([52, 65, 32, 39, 32, 35, 69, 26, 31, 45, 33, 42, 66, 27, 19, 24, 43,
       30, 68, 19, 33, 58, 31, 50, 21, 70, 52, 62, 55, 74, 22, 63, 21, 42,
       32, 71, 68, 31, 32, 72, 56, 23, 56, 38, 74, 52, 67, 21, 52, 52, 22,
       65, 46, 46, 63, 36, 28, 57, 54, 54, 56, 22, 36, 62, 57, 38, 41, 40,
       55, 29, 62, 72, 65, 60, 37, 71, 32, 30, 58, 50, 27, 59, 26, 52, 61,
       41, 50, 35, 40, 68, 22, 52, 62, 43, 69, 44, 54, 25, 42, 69])

In [64]:
# Let's find out the mean of age_sample2
print(ages_sample2.mean())

58.6


In [65]:
# LEt's import the library

from scipy.stats import ttest_ind

In [66]:
# now let's find out the ttest

ttest_ind(a=sample_age,b=ages_sample2,equal_var=False) # equal_var assumes samples have equal variance or not, we consider it as no different variance

Ttest_indResult(statistic=-7.0370809595027275, pvalue=1.0236096529996912e-10)


#### Hence the ALTERNATE hypothesis is True which means both the samples belong to different distributions

## Paired T Test

#### Now let's discuss about a type of T test which deals with the samples of same population at different time intervals.

#### For example let's say a particular vaccine is given a particlular group of people and we want to check if it has any impact on their weights measured. We have taken their weights before the vaccination and after the vaccination and now we want to check is there any difference in the weights beofre and after the vaccination

#### So here we do the Paired T Test

In [60]:
np.random.seed(15)

before_vaccination=stats.norm.rvs(scale=30,loc=250,size=100)

after_vaccination=before_vaccination+stats.norm.rvs(scale=5,loc=1.25,size=100)

weights=pd.DataFrame({'before':before_vaccination,
                     'after':after_vaccination,
                     'difference':after_vaccination-before_vaccination})

weights

before       after  difference
0   240.630146  235.459622   -5.170523
1   260.178541  258.357256   -1.821286
2   245.322744  246.033397    0.710653
3   234.946310  241.864426    6.918116
4   257.067067  253.135835   -3.931231
..         ...         ...         ...
95  258.590998  264.748878    6.157880
96  218.560838  214.624371   -3.936467
97  244.406493  248.025206    3.618713
98  264.926647  264.080516   -0.846131
99  237.864934  245.567503    7.702569

[100 rows x 3 columns]

In [61]:
# So here we can use the ttest_rel function

from scipy.stats import ttest_rel

In [62]:
ttest_rel(a=before_vaccination,b=after_vaccination)

Ttest_relResult(statistic=-1.9084762985889763, pvalue=0.059226433816612)

#### Here the pvalue is 0.059226433816612 which is greater than the significance value 0.05, hence we can conclude that weight does not has an impact on the vaccination

### Now this sample data is being prepared based pure imagination and has no practical evidence